ledger/participant-integration-api/src/main/scala/platform/index/IndexServiceImpl.scala

```diff
@@ -309,31 +293,85 @@ private[index] final class LedgerBackedIndexService(
   ): Source[(domain.LedgerOffset.Absolute, domain.ConfigurationEntry), NotUsed] =
     Source
       .future(concreteOffset(startExclusive))
-      .flatMapConcat(ledger.configurationEntries(_).map { case (offset, config) =>
-        toAbsolute(offset) -> config.toDomain
+      .flatMapConcat(dispatcher.startingAt(_, RangeSource(ledgerDao.getConfigurationEntries)).map {
+        case (offset, config) =>
+          toAbsolute(offset) -> config.toDomain
       })
 
-  /** Participant pruning command */
   override def prune(pruneUpToInclusive: Offset, pruneAllDivulgedContracts: Boolean)(implicit
       loggingContext: LoggingContext
-  ): Future[Unit] =
-    ledger.prune(pruneUpToInclusive, pruneAllDivulgedContracts)
-
-  private def concreteOffset(startExclusive: Option[LedgerOffset.Absolute]): Future[Offset] =
-    startExclusive
-      .map(off => Future.fromTry(ApiOffset.fromString(off.value)))
-      .getOrElse(Future.successful(Offset.beforeBegin))
+  ): Future[Unit] = {
+    pruneBuffers(pruneUpToInclusive)
+    ledgerDao.prune(pruneUpToInclusive, pruneAllDivulgedContracts)
+  }
 
   override def getMeteringReportData(
       from: Timestamp,
       to: Option[Timestamp],
       applicationId: Option[ApplicationId],
-  )(implicit loggingContext: LoggingContext): Future[ReportData] = {
-    ledger.meteringReportData(
+  )(implicit loggingContext: LoggingContext): Future[ReportData] =
+    ledgerDao.meteringReportData(
       from: Timestamp,
       to: Option[Timestamp],
       applicationId: Option[ApplicationId],
     )
+
+  override def currentLedgerEnd()(implicit
+      loggingContext: LoggingContext
+  ): Future[LedgerOffset.Absolute] = {
+    val offset =
+      if (ledgerEnd() == Offset.beforeBegin) ApiOffset.begin
+      else ledgerEnd()
+    Future.successful(toAbsolute(offset))
   }
 
+  private def ledgerEnd(): Offset = dispatcher.getHead()
```

Comment:
`LedgerEndCache` doesn't have any description, what guarantees does it give? Is it possible that the ledger end returned by `LedgerEndCache` is ahead or behind the one broadcast by the dispatcher? If so, can it affect the correctness of the ledger api server?

Reply:
> LedgerEndCache doesn't have any description, what guarantees does it give? Is it possible that the ledger end returned by LedgerEndCache is ahead or behind the one broadcast by the dispatcher? If so, can it affect the correctness of the ledger api server?

It needs to be ahead of the dispatcher since it constrains the underlying SQL queries (I faced some immediate pain in conformance tests as implementing it in the other way around)
Funny thing is: right now it is two synchronous calls after each other ;)